decoder/context.go

```diff
@@ -16,8 +21,79 @@ type DecoderContext struct {
 	// CodeLenses represents a slice of executable lenses
 	// which will be executed in the exact order they're declared
 	CodeLenses []lang.CodeLensFunc
+
+	CompletionHooks        CompletionFuncMap
+	CompletionResolveHooks CompletionResolveFuncMap
 }
 
 func (d *Decoder) SetContext(ctx DecoderContext) {
 	d.ctx = ctx
 }
+
+type CompletionFunc func(ctx context.Context, value cty.Value) ([]Candidate, error)
+type CompletionFuncMap map[string]CompletionFunc
+
+type CompletionResolveFuncMap map[string]CompletionResolveFunc
+type CompletionResolveFunc func(ctx context.Context, unresolvedCandidate UnresolvedCandidate) (*ResolvedCandidate, error)
+
+type Candidate struct {
+	// Label represents a human-readable name of the candidate
+	// if one exists (otherwise Value is used)
+	Label string
+
+	// Detail represents a human-readable string with additional
+	// information about this candidate, like symbol information.
+	Detail string
+
+	Kind lang.CandidateKind
+
+	// Description represents human-readable description
+	// of the candidate
+	Description lang.MarkupContent
+
+	// IsDeprecated indicates whether the candidate is deprecated
+	IsDeprecated bool
+
+	RawInsertText string
+
+	// ResolveHook represents a resolve hook to call
+	// and any arguments to pass to it
+	ResolveHook *lang.ResolveHook
+}
+
+type ExpressionCandidate struct {
```

Comment:
Do you mind adding a comment to this, to indicate that this is the preferred way to provide completion candidates for attributes values (expressions) and to note that there's `ExpressionCompletionCandidate` to convert it into the low-level `Candidate`?